### DISCUSSION TITLE
Representation of ozone retrieved from prepbufr files and stored into netcdf files

### DISCUSSION THREAD

**Discussion by edwardstrobach** [dtcenter/METplus#1951] [Statistical Computation] [answered]
Hello,

I used the pb2nc step to create netcdf files that store ozone across CONUS for each day.  I am appending the meta data structure below.  I have been using a subset of the data for evaluations and have noticed some interesting temporal variability that is inconsistent with the surface ozone stations posted in a separate screenshot below during the same time of this file.  Note that the variability I am refer to which is shown in the plot attached is widespread across the region and is evident during the entire month of August 2021 for the LA basin area, so its not an isolated incident spatially or temporally. 

**>>> dataset
<class 'netCDF4._netCDF4.Dataset'>
root group (NETCDF4 data model, file format HDF5):
    MET_Obs_version: 1.02
    use_var_id: true
    FileOrigins: File / generated 20220214_132744 UTC on host h16c46 by the MET pb2nc tool
    MET_version: V9.1.1
    MET_tool: pb2nc
    dimensions(sizes): mxstr(16), mxstr2(40), mxstr3(80), nobs(55966), nhdr(55966), nhdr_typ(1), nhdr_sid                                                     (1188), nhdr_vld(24), nobs_qty(2), obs_var_num(1)
    variables(dimensions): int32 hdr_typ(nhdr), int32 hdr_sid(nhdr), int32 hdr_vld(nhdr), float32 hdr_lat                                                     (nhdr), float32 hdr_lon(nhdr), float32 hdr_elv(nhdr), int32 obs_qty(nobs), int32 obs_hid(nobs), int32 obs                                                     _vid(nobs), float32 obs_lvl(nobs), float32 obs_hgt(nobs), float32 obs_val(nobs), |S1 hdr_typ_table(nhdr_t                                                     yp, mxstr2), |S1 hdr_sid_table(nhdr_sid, mxstr2), |S1 hdr_vld_table(nhdr_vld, mxstr), |S1 obs_qty_table(n                                                     obs_qty, mxstr), |S1 obs_var(obs_var_num, mxstr2), |S1 obs_unit(obs_var_num, mxstr2), |S1 obs_desc(obs_va                                                     r_num, mxstr3)**

1.  I separated the station lat/lon for the site of interest using the example below:

**>>> lon[i0[i1]]
masked_array(data=[-118.23, -118.23, -118.23, -118.23, -118.23, -118.23,
                   -118.23, -118.23, -118.23, -118.23, -118.23, -118.23,
                   -118.23, -118.23, -118.23, -118.23, -118.23, -118.23,
                   -118.23, -118.23, -118.23, -118.23, -118.23, -118.23,
                   -118.23, -118.23, -118.23, -118.23, -118.23, -118.23,
                   -118.23, -118.23, -118.23, -118.23, -118.23, -118.23,
                   -118.23, -118.23, -118.23, -118.23, -118.23, -118.23,
                   -118.23, -118.23, -118.23, -118.23, -118.23, -118.23],
             mask=False,
       fill_value=1e+20,
            dtype=float32)
>>> lat[i0[i1]]
masked_array(data=[34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07,
                   34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07,
                   34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07,
                   34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07,
                   34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07,
                   34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07, 34.07],
             mask=False,
       fill_value=1e+20,
            dtype=float32)**

The file consists of 48 records during the 24-hr period.  Each file stores ozone information every hour each day.  The reason for 48 records is because there is 1-hr accumulation time and an 8-hr accumulation time.

**>>> obs_lvl[i0[i1]].data/3600
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.,
       8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.],
      dtype=float32)**

There are 48 records here too. 

2.  From there, I separated out the accumulation times and produced a plot below.  One is for a 1-hr accumulation time while the other is for an 8-hr accumulation time.  Both produce quite a bit of variability in the temporal structure.  The peaks associated with 8-hr accumulations are typically more pronounced but not always.  However, as you can see, choosing one over the other, averaging between the two, or selecting data that yields the smallest time difference would not remove the variability that we were seeing. I would think that sticking to 1-hr accumulations is most appropriate. However, if you compare the plot below with the link, you can see that the ozone representation within the link (local time) is temporally more gradual.  I am preparing a talk for AGU and I want to make sure what I'm using is reliable or if there is something that I'm not considering or doing incorrectly.  Thank you in advance. 

<img width="576" alt="Screen Shot 2022-11-16 at 4 15 11 PM" src="https://user-images.githubusercontent.com/48096468/202299803-7e0b59e5-eba4-448b-a640-c48b19526ed3.png">

<img width="753" alt="Screen Shot 2022-11-16 at 4 42 49 PM" src="https://user-images.githubusercontent.com/48096468/202300863-9362d644-e8fe-4bb6-b727-9e52e9f3dc1f.png">

**Comment by CPKalb** [comment]
Hi Edward:

I want to clarify, to make sure I understand what you are asking.  It sounds like you are comparing the two plots at the bottom of the question?  The first plot is output from MET's conversion of pb2nc, while the second is data that is from a link?  Is this correct?  My confusion is that the two plots seem to be on very different y-axis scales as one ranges from 0 to 0.02, while the other ranges from 0 to 60.  Should these variables look the same (and they are very different in terms of values), or is the concern that the top graphic has many more pronounced peaks than the bottom graphic?

Thanks,
Christina

**Comment by edwardstrobach** [comment]
Hi Christina,

Thank you for the quick response.  I had used metplus to convert the units
to ppb and used the results from the output to generate the plots showing
the two curves with different time accumulations (1-hr v. 8-hr) while the
plot below that (second attachment) is from a website with units in ppm, so
multiplying by 1000 should get you a similar scale.  A colleague of mine
took all the stations for the day of interest from that website and did not
see the variations I'm seeing in the plot that I generated using output
from the pb2nc step.

**Reply by CPKalb** [reply]
Hi Edward:

Thanks.  If I multiply the bottom plot values by 1000, I get a plot range of 0 to 200.  The top plot then has a scale of 0 to 60.  Variability is always magnified when the y axis range is shortened.  Would it be possible to get the top graphic to be on the same scale (0 to 200) as the bottom?  This will make it much easier to see how the variability differs between the two datasets.

Thanks,
Christina

**Comment by edwardstrobach** [comment]
Sure, I redid the plot with the scale going from 0-200ppb.   Since the plot
online is in PST and the plot attached below is in utc, then I decided to
also mark on plots 4p, 4a, and 4p to denote 00z, 12z, and 00z (start of
next day).  Normally one would see a ramp up in ozone during the day due to
photochemistry; that's not really happening here.  Thank you again.

**Reply by CPKalb** [reply]
Hi Edward:

Could you post the plot here for comparison purposes?  Also, it would be helpful to see your pb2nc configuration file, as well as how you are running the tool.  For example, are you running from a metplus configuration file or from a command line?  What is the input frequency of your data?  If it's possible to upload a data file from a problem time period, that would be helpful.  Instructions to upload data can be found [here](https://github.com/dtcenter/METplus/discussions/954).

Thanks,
Christina

**Reply by edwardstrobach** [reply]
No problem.  The attached plot is below.  Looking at my old bash scripts that run metplus, it would appear that the specific file that is used to produce the netcdf files that contain the airnow data is called aqm.t12z.prepbufr.tm00.  I just pulled down that file from hera and attempted to begin transfer the prepbufr data, the netcdf file generated during pb2nc, and both configuration files over but was denied access to directories within the ftp site under passive mode after entering my noaa email.  The two configuration files that I'm referring to are referenced in the bash that script and are called pb2nc_aq.conf, and within that config file is a reference to PB2NCConfig_aqm.  As far as I can tell the data is hourly.  

Any idea how I can get permission to upload these four files?

<img width="581" alt="o3_08052021" src="https://user-images.githubusercontent.com/48096468/202448978-cbe55bec-a4c1-42fc-88ce-bc4231379a06.png">

**Reply by CPKalb** [reply]
@jprestop do you have any idea why someone would be denied access to the ftp after entering their noaa email?

**Reply by CPKalb** [reply]
Hi Edward:

I'll also need some information about the data in the comparison plot.   Specifically, how are these plots created.  Is the data raw or is there any post processing done on it.  If there is a website that may have that information, a link would be useful as well

Tina

**Reply by georgemccabe** [reply]
@edwardstrobach, did you enter `anonymous` as the username and your email address as the password?

**Comment by edwardstrobach** [comment]
Hello Christina and George,

I did follow the exact instructions with username and password as was
referenced in the email but was denied access after entering the "
***@***.***" password.  The plot that is generated was done
using python using the netcdf file generated from pb2nc.  In the file I
isolated a single site as a test within the area of interest and retrieved
48 records, one for a time accumulation of an hour (24 records) and another
for a time accumulation of 8 hours (24 records).  I further isolated those
sets of records, time ordered the two sets and plotted the result that is
shared in the previous post.  Thanks!

**Reply by CPKalb** [reply]
Hi Edward:

Thanks.  I also need information about how the web plot was computed.  Is there a link or anything that would give more information about whether there is any pre-processing done on that image.

Thanks,
Christina

**Reply by georgemccabe** [reply]
That is odd about being denied access to the FTP. I have not seen that issue before. I was able to enter my NOAA email address and get in without issue. Perhaps there is an issue connecting from the computer you are using. You could try transferring the files to another machine and try to upload them to the FTP from there?

**Reply by CPKalb** [reply]
Hi Edward:

You can likely attach the configuration files here if you add/change the file extension to be .txt.  If you have a chance to try uploading the prepbufr files from another computer, that would be helpful.

Christina

**Comment by edwardstrobach** [comment, marked as answer]
I'm still waiting to hear back from a colleague about this data.  I did see
this on the website, however:  Data after 2020 may be preliminary. AQMIS
data are in PST.

I'm not sure if that suggests a lack in preprocessing or if the data is a
less processed version.  I see that I can also download .csv files for
groupings of sites within an area for each hour or for simply just one
site.  Time period selection also seems flexible.  I don't think that will
address your preprocessing question though.

I believe my personal macbook is causing the issue as George pointed out.
I'm not sure if a vpn would work.  I have made gains on file transfers from
my NOAA work laptop.  However, I still experience an issue when running
the  'put "my files"' command.  I thought that inputting the following line
might work:  put "/Users/estrobach/Desktop/prepbufr.aqm.20210805.nc" would
work, and although is said "200 PORT command successful", it seems to have
failed at initiating file transfer as indicated by "425 Unable to build
data connection: Connection timed out".  The file tested was only 173kB.

All these files are on google drive right now.  I would like to know if I
did something wrong with attempting a file transfer.

Thanks

**Reply by jprestop** [reply]
Hi @edwardstrobach.  I have been out of the office and just now saw @CPKalb's comment:

> @jprestop do you have any idea why someone would be denied access to the ftp after entering their noaa email?
There should be no problem with an noaa email.  

It doesn't look like you did anything incorrectly.  Perhaps there was a problem with our ftp server at that time.  It's hard to say.  I would recommend either trying again, or as @CPKalb suggested, attaching the configuration files here if you add/change the file extension to be .txt.

**Reply by jprestop** [reply]
Hi @edwardstrobach.  I just wanted to follow up and see if you still needed any assistance with this issue?

**Reply by jprestop** [reply]
Hi @edwardstrobach.  I just wanted to give you an opportunity to follow up and/or select an answer for this Discussion.  If we don't hear back in the next week or so, we'll go ahead and select an answer and lock this Discussion.
